R/ddl.R

```diff
@@ -0,0 +1,193 @@
+#' DDL object
+#'
+#' Object to execute custom DDL code in the shiny session
+#'
+#' @param expr (`expression`)\cr
+#'  Syntatically valid R code to be executed in the shiny session.
+#'  shouldn't be specified when `code` is specified.
+#'
+#' @param code (`character`, `language`)\cr
+#'  Object containing code to be evaluated to load data. Shouldn't be specified when `expr`
+#'  is specified.
+#'
+#'
+#' @param ui (`shiny.tag`)\cr
+#'   `shiny` ui module containing inputs which `id` correspond to the
+#'   args in the `code`.
+#'
+#' @param server (`function`)\cr
+#'   `shiny` server module returning data. This server should execute
+#'   `code` and return a reactive data containing necessary data. To handle
+#'   evaluation and code masking process it is recommended to use `ddl_run`.
+#'   Package provides universal `username_password_server` which
+#'   runs [ddl_run] function, which returns `teal_data` object.
+#'   Details in the the example
+#'
+#' @param input_mask (`list` named)\cr
+#'   arguments to be substituted in the `code`. These
+#'   argument are going to replace arguments set through
+#'   `ui` and `server`. Example use case is when app user
+#'   is asked to input a password and we'd like to skip this
+#'   input in the reproducible code. Typically users password
+#'   is substituted with `askpass::askpass()` call, so the
+#'   returned code is still executable but secure.
+#'
+#' @param datanames (`character`)\cr
+#'   Names of the objects to be created from the code evaluation.
+#'   If not specified (`character(0)`), all objects will be used to `teal_data` function
+#'   (via `env_list` in `postprocess_fun`).
+#'
+#' @inheritParams teal.data::teal_data
+#'
+#'
+#' @export
+ddl <- function(expr,
+                code,
+                ui = submit_button_ui,
+                input_mask = list(),
+                server = submit_button_server,
+                join_keys = teal.data::join_keys(),
+                datanames) {
+  if (!missing(expr) && !missing(code)) {
+    stop("Only one of `expr` or `code` should be specified")
+  }
```

Comment:
[`rlang::check_exclusive`](https://rlang.r-lib.org/reference/check_exclusive.html)

(please implement in other places if necessary)

Reply:
I'm not particularly familiar with `rlang` error messages myself.

Reply:
rlang error system ftw!

Reply:
One of the main features of `rlang@1.0.0` is better display of the error messages - [link](https://www.tidyverse.org/blog/2021/12/rlang-1-0-0-errors/). Since the release of this version, most (if not all?) of tidyverse packages followed this. That's why I do believe that this might be more coherent across all the packages one might use hence: easier to understand and take appropriate action.

Reply:
I had issues in the past year with `rlang` handling `errors`. Needed to use `rlang::cnd_message_lines(error)` on Windows on every `error` message in `tryCatch` to get an accurate `error` message. Otherwise errors were always empty. If people have Windows and old `rlang` versions they may not see an error at all.

Reply:
I would also vote up for striving away from `tidyverse` as they have a great history of breaking changes and tend to have multiple dependencies in it. I don't think general error messages in `rlang` are better than customized, personalized human-crafted ones that we thought through to inform the user in the most useful manner as possible.

Reply:
I do echo similar mixed feeling about tidyverse but this is actually about r-lib which I consider much more solid.

> I don't think general error messages in rlang are better than customized, personalized human-crafted ones that we thought through to inform the user in the most useful manner as possible.

Yes and no. Yes because the role of the error message is to clearly indicate what's wrong and why to allow a quick fix by the end users. No, because if we will be too verbose then it won't be clear anymore. The other thing that I am actually trying to address here is consistency. Just a few examples of our codebase that I managed to find within few minutes:
https://github.com/insightsengineering/teal.slice/blob/686f1f76ac6130325df9314fea419800c139f35a/R/teal_slice.R#L134-L135
https://github.com/insightsengineering/teal.modules.general/blob/eef36c1dda918fcbebe948d8bef189bd7e545974/R/tm_g_bivariate.R#L762-L764
https://github.com/insightsengineering/teal.data/blob/c04b7edef44d80fd4cdf0c6196d5c9c3de7c84b1/R/utils.R#L21-L24
https://github.com/insightsengineering/teal.slice/blob/686f1f76ac6130325df9314fea419800c139f35a/R/teal_slice.R#L134-L135
If we can delegate the check as well as error message for a separate functionality called from all of those places then we would achieve some consistency. It's mostly done via `checkmate` for individual objects checks. I hope that `rlang` is a good candidate to fill in the gaps of what cannot be done via `checkmate`.
One might raise that we don't want to introduce a new dependency but `rlang` is very likely to be already there (directly or indirectly). What's great about it is that this is a zero-dependency package. That changed my view on this package quite significantly.

Reply:
Those 4 examples are actually fine in my opinion. Very clear :)
If rlang has zero dependencies than I'm less against, but still it's better to have a package that has 8 dependencies, than a packages with 15 dependencies where 7 of them has zero dependencies. The overview of such package become repellent. Each package extends the time of the installation. And each package require us and some more advanced users who would like to read the code to understand another workflow. How many workflows can we use in a teal designed to be simple and seamless for R users?